### ISSUE TITLE
YOLOv5 issues with `torch==1.12` on Multi-GPU systems

### ISSUE THREAD

**Issue by glenn-jocher** [ultralytics/yolov5#8395]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Multi-GPU

### Bug

All GPUs are utilized by torch 1.12 with current YOLOv5 master when a single-GPU command is run, i.e.:

Error does not occur with torch==1.11

@AyushExel FYI

### Environment

Docker image

### Minimal Reproducible Example

```
python train.py --epochs 10 --device 7
```

### Additional

Temp workaround is to use torch 1.11
```
pip install torch==1.11.0+cu113 torchvision==0.12.0+cu113 torchaudio==0.11.0 --extra-index-url https://download.pytorch.org/whl/cu113
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by AyushExel** [comment]
I wonder how a bug this big made into the release

**Comment by UnglvKitDe** [comment]
@glenn-jocher I cannot confirm this. Tested on gpu servers with 2,4,8 GPUS (RTX 2080TI/A6000)
EDIT: Maby a workaround is to use CUDA_VISIBLE_DEVICES?

**Comment by glenn-jocher** [comment]
@AyushExel it's probably not a torch bug but instead related to our specific implementation for selecting devices in select_device() below which relies on defining CUDA_VISIBLE_DEVICES in the workspace before torch reads it.
mode.https://github.com/ultralytics/yolov5/blob/898332433a71b8846b15daa276a8ac45c9efa98b/utils/torch_utils.py#L52-L86

If I run the reproduce command above in Docker image, `python train.py --epochs 10 --device 7`, I see this on our server:

<img width="871" alt="Screen Shot 2022-06-30 at 4 28 13 PM" src="https://user-images.githubusercontent.com/26833433/176703558-827ee81d-a664-4a8c-b5e5-8cc3c9a2ffdc.png">

**Comment by glenn-jocher** [comment]
@UnglvKitDe you tested a single-GPU training command above `python train.py --epochs 10 --device 7` with torch 1.12 and did not see multi-GPU usage on nvidia-smi?

**Comment by UnglvKitDe** [comment]
@glenn-jocher 

> @UnglvKitDe you tested a single-GPU training command above `python train.py --epochs 10 --device 7` with torch 1.12 and did not see multi-GPU usage on nvidia-smi?

@glenn-jocher not exactly, but in the end the same. i tested it on coco128 ( torch 1.12 and cuda 11.6). so `python train.py --data coco128.yaml --device 0 --epochs 10`.

**Comment by glenn-jocher** [comment]
@UnglvKitDe oh, strange. This is basically the same command I used, but I used device 7. If I try --device 0 I also get the bug though.

We'll ok I'll experiment some more. In my experiments 1.11 works correctly but 1.12 does not (master branch in Docker) with CUDA 11.3

**Comment by mjun0812** [comment]
It appears that this problem occurred because of a change in the timing of reading environment variables in PyTorch 1.12.

My GPU environment is here.

```bash
CUDA:0 NVIDIA RTX A6000, 48685.3125MB
CUDA:1 NVIDIA GeForce RTX 3090, 24268.3125MB
```

The following code worked correctly in version 1.11.

```python
import os
import torch

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

In version 1.11, the output of this code is below.

```bash
Using GPU is CUDA:1
CUDA:0 NVIDIA GeForce RTX 3090, 24268.3125MB
```

But version 1.12's output is below.

```bash
Using GPU is CUDA:1
CUDA:0 NVIDIA RTX A6000, 48685.3125MB
CUDA:1 NVIDIA GeForce RTX 3090, 24268.3125MB
```

As additional information, the environment variable change worked correctly when done before `import torch`.

```python
import os

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

import torch


# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

Output is below.

```bash
Using GPU is CUDA:1
CUDA:0 NVIDIA GeForce RTX 3090, 24268.3125MB
```

**Comment by glenn-jocher** [comment]
@mjun0812 interesting, thanks for the info! Unfortunately we can't specify the environment variables before loading torch. Does re-loading torch after defining the environment variables have any effect? i.e.:

```python
import os
import torch

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

import torch


# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

**Comment by mjun0812** [comment]
@glenn-jocher Thank you for your reply!
I tried your code and the following code.

Your suggestion code: 

```python
import os
import torch

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

import torch

# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

and `del torch`

```python
import os
import torch

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

del torch
import torch

# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

and `importlib.reload()`

```python
import os
import torch
import importlib

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

importlib.reload(torch)

# print using GPU Info
print(f"Using GPU is CUDA:{os.environ['CUDA_VISIBLE_DEVICES']}")

for i in range(torch.cuda.device_count()):
    info = torch.cuda.get_device_properties(i)
    print(f"CUDA:{i} {info.name}, {info.total_memory / 1024 ** 2}MB")
```

The above outputs were the same.

```bash
Using GPU is CUDA:1
CUDA:0 NVIDIA RTX A6000, 48685.3125MB
CUDA:1 NVIDIA GeForce RTX 3090, 24268.3125MB
```

This issue could not be resolved...
I will continue to investigate the issue.

**Comment by glenn-jocher** [comment]
@mjun0812 too bad. Yes let me know if you find a solution!

**Comment by mjun0812** [comment]
@glenn-jocher I raised this [issue](https://github.com/pytorch/pytorch/issues/80876) in the PyTorch repository, and it has been [fixed](https://github.com/pytorch/pytorch/pull/80899) in the latest master branch.
The cause is that CUDA initialization was done during `import torch`.
The binary version that can be installed with pip will likely be fixed in the torch next version 1.12.1.

Therefore, it may be necessary to modify `requirements.txt` as follows

https://github.com/ultralytics/yolov5/blob/fdc9d9198e0dea90d0536f63b6408b97b1399cc1/requirements.txt#L12-L13

```diff
+ torch>=1.7.0,!=1.12.0  # https://github.com/ultralytics/yolov5/issues/8395
+ torchvision>=0.8.1,!=0.13.0  # https://github.com/ultralytics/yolov5/issues/8395
```

I am ready to make a pull request for the above fixes.

**Comment by glenn-jocher** [comment]
@mjun0812 got it, thanks for the update! I see your PR, will take a look there.

**Comment by glenn-jocher** [comment]
@AyushExel torch 1.12 issue resolved in upcoming torch 1.12.1, so our fix is to simply excluded 1.12 in requirements.txt in #8497. Problem solved :)

**Comment by DoSquared** [comment]
How did you guys fix the pytorch error? should I install 1.12.1?

**Comment by glenn-jocher** [comment]
@DaliaMahdy unfortunately 1.12.1 is not out currently, latest stable is 1.12.0, but you can install nightly for example to resolve the issue, or simply use 1.11.0
